File: mobile/lib/utils/inbox.dart
import '../models/customer.dart';
import '../models/message.dart';
import '../utils/phone.dart';

class InboxThread {
  final String phoneNumber;
  final String lastBody;
  final DateTime lastAt;
  final String direction;
  final Customer? customer;

  InboxThread({
    required this.phoneNumber,
    required this.lastBody,
    required this.lastAt,
    required this.direction,
    this.customer,
  });
}

List<InboxThread> buildThreads(List<Message> messages, List<Customer> customers) {
  final byPhone = <String, List<Message>>{};
  for (final message in messages) {
    final key = phoneKey(message.phoneNumber);
    byPhone.putIfAbsent(key, () => []).add(message);
  }

  final threads = <InboxThread>[];
  for (final entry in byPhone.entries) {
    final sorted = [...entry.value]
      ..sort((a, b) => b.createdAt.compareTo(a.createdAt));
    final last = sorted.first;
    Customer? customer;
    for (final c in customers) {
      if (phoneKey(c.phoneNumber) == entry.key) {
        customer = c;
        break;
      }
    }
    threads.add(
      InboxThread(
        phoneNumber: entry.key,
        lastBody: last.body,
        lastAt: last.createdAt,
        direction: last.direction,
        customer: customer,
      ),
    );
  }

  threads.sort((a, b) => b.lastAt.compareTo(a.lastAt));
  return threads;
}

File: lib/inbox.ts
import { phoneKey } from "@/lib/phone-links";
import type { Customer, InboxThread, Message } from "@/lib/types";

export function buildThreads(
  messages: Message[],
  customers: Pick<Customer, "id" | "name" | "phone_number">[],
): InboxThread[] {
  const byPhone = new Map<string, Message[]>();
  for (const message of messages) {
    const key = phoneKey(message.phone_number);
    const list = byPhone.get(key) ?? [];
    list.push(message);
    byPhone.set(key, list);
  }

  const threads: InboxThread[] = [];
  for (const [phone_number, list] of byPhone) {
    const sorted = [...list].sort(
      (a, b) => new Date(b.created_at).getTime() - new Date(a.created_at).getTime(),
    );
    const last = sorted[0];
    const customer =
      customers.find((row) => phoneKey(row.phone_number) === phone_number) ?? null;
    threads.push({
      phone_number,
      last_body: last.body,
      last_at: last.created_at,
      direction: last.direction,
      customer,
    });
  }

  return threads.sort(
    (a, b) => new Date(b.last_at).getTime() - new Date(a.last_at).getTime(),
  );
}
